xapian-core/tests/api_queryparser.cc

```diff
@@ -2745,6 +2745,11 @@ qp_scale1_helper(const Xapian::Database &db, const string & q, unsigned n,
 // Regression test: check that query parser doesn't scale very badly with the
 // size of the query.
 DEFINE_TESTCASE(qp_scale1, synonyms) {
+    const char* xfail_msg = "Querying synonyms for multiple query words is "
+	               "expected to incur a higher time in remote DB";
+    XFAIL_FOR_BACKEND("remote", xfail_msg);
+    XFAIL_FOR_BACKEND("multi_remoteprog_glass", xfail_msg);
+    XFAIL_FOR_BACKEND("multi_glass_remoteprog_glass", xfail_msg);
```

Comment:
I don't think we want to `XFAIL` here - that means an `XPASS` result (which means "unexpected pass" and the testcase is treated as failing) if the time is OK and that's what I seem to reliably get:

```
$ ./runtest ./apitest qp_scale1 -v
Running test './apitest qp_scale1 -v' under eatmydata and valgrind
Running tests with backend "honey"...
Running tests with backend "none"...
Running tests with backend "inmemory"...
Running tests with backend "glass"...
Running test: qp_scale1... POSSIBLE UNRELEASED MEMORY - RETRYING TEST ok
./apitest backend glass: All 1 tests passed.
Running tests with backend "singlefile_glass"...
Running tests with backend "multi_glass"...
Running test: qp_scale1... ok
./apitest backend multi_glass: All 1 tests passed.
Running tests with backend "multi_glass_remoteprog_glass"...
Running test: qp_scale1... POSSIBLE UNRELEASED MEMORY - RETRYING TEST XPASS (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
Running tests with backend "multi_remoteprog_glass"...
Running test: qp_scale1... POSSIBLE UNRELEASED MEMORY - RETRYING TEST XPASS (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
Running tests with backend "remoteprog_glass"...
Running test: qp_scale1... XPASS (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
Running tests with backend "remotetcp_glass"...
Running test: qp_scale1... XPASS (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
./apitest total: 2 tests passed, 4 expected failures passed.
```

That's with your branch rebased onto current git master, but I don't think there are any changes on git master which would have affected this.

Probably just making the condition `synonyms && !remote` will do for now (unless we manage to solve this before merging).

If you run `./runtest ./apitest -v qp_scale1` without these, what's the output?  I'm wondering which part of the testcase is actually failing.  I tried printing out the times and most of the ratios of large/small are < 1 for me.

Reply:
If I suppress use of valgrind then I mostly seem to get XFAIL, with the occasional XPASS:

```
$ VALGRIND= ./runtest ./apitest qp_scale1 -v
Running test './apitest qp_scale1 -v' under eatmydata
Running tests with backend "honey"...
Running tests with backend "none"...
Running tests with backend "inmemory"...
Running tests with backend "glass"...
Running test: qp_scale1... ok
./apitest backend glass: All 1 tests passed.
Running tests with backend "singlefile_glass"...
Running tests with backend "multi_glass"...
Running test: qp_scale1... ok
./apitest backend multi_glass: All 1 tests passed.
Running tests with backend "multi_glass_remoteprog_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
small=0.010077s, large=0.004406s
small=0.048833s, large=0.095696s
small=0.056213s, large=0.150561s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.150561 < 0.120858


Running tests with backend "multi_remoteprog_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
small=0.00565s, large=0.004249s
small=0.101363s, large=0.193138s
small=0.105056s, large=0.311184s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.311184 < 0.22587


Running tests with backend "remoteprog_glass"...
Running test: qp_scale1... XPASS (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
small=0.009589s, large=0.003246s
small=0.052061s, large=0.082922s
small=0.065391s, large=0.117193s
small=0.058717s, large=0.03465s

Running tests with backend "remotetcp_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher time in remote DB)
small=0.007938s, large=0.004389s
small=0.085934s, large=0.151948s
small=0.113013s, large=0.290744s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.290744 < 0.242978


./apitest total: 2 tests passed, 1 expected failures passed, 3 expected failures.
```

So for me it seems to always be the third part that fails - that's parsing repetitions of `foo` with a synonym `foo` -> `bar` and no multi-word synonyms (but with `FLAG_AUTO_MULTIWORD_SYNONYMS` on).

If I switch that part to using `FLAG_AUTO_SYNONYMS` then I seem to reliably get `XPASS`, so it seems that checking for but not finding multi-word synonyms isn't scaling well - that's something I should be able to investigate.

The second sub-testcase seems to reliably take longer in the larger version, but that changes to usually being quicker with just `FLAG_AUTO_SYNONYMS`.  The fourth sub-testcase is reliably quicker in the larger version though, and that actually uses multi-word synonyms.  That reinforces that it's probably not finding multi-word synonyms that's the issue here.

Reply:
Third one consistently fails for me aswell:
```
Running tests with backend "multi_glass_remoteprog_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher cost in remote DB)
small=0.010694s, large=0.009417s
small=0.040276s, large=0.056852s
small=0.030887s, large=0.083621s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.083621 < 0.066407


Running tests with backend "multi_remoteprog_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher cost in remote DB)
small=0.011954s, large=0.010813s
small=0.085493s, large=0.102148s
small=0.054816s, large=0.155688s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.155688 < 0.117854


Running tests with backend "remoteprog_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher cost in remote DB)
small=0.003145s, large=0.002939s
small=0.036848s, large=0.058566s
small=0.039391s, large=0.091513s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.091513 < 0.0846907


Running tests with backend "remotetcp_glass"...
Running test: qp_scale1... XFAIL (Querying synonyms for multiple query words is expected to incur a higher cost in remote DB)
small=0.010367s, large=0.009933s
small=0.073369s, large=0.088867s
small=0.052924s, large=0.132596s
api_queryparser.cc:2742: (time2) < (time1 * 2.15)
Evaluates to: 0.132596 < 0.113787
```

Reply:
Is there a documentation for XFAIL?
I was trying to look for it but couldn't find. I think I tried developer guide and few other places. I was bit confused on usage of XFAIL myself.

Reply:
Yes, removing `FLAG_AUTO_SYNONYMS` seems to fix the test case and return `XPASS`. I will remove XFAIL part for now and probably it's worth wait if it's possible for you to investigate this.

Reply:
> Is there a documentation for XFAIL?
> I was trying to look for it but couldn't find. I think I tried developer guide and few other places. I was bit confused on usage of XFAIL myself.

There's a note which covers how it works and a common use case here: https://xapian-developer-guide.readthedocs.io/en/latest/contributing/contributing-changes.html#automated-tests

There's also a documentation comment where it's defined in the code: xapian-core/tests/harness/testsuite.h